File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    p:=head
    m:=make(map[*Node]*Node)
    dummyNode:=new(Node)
    dummyNode.Val=-1
    temp:=dummyNode
    for p!=nil{
        temp.Next=new(Node)
        temp=temp.Next
        temp.Val=p.Val
        m[p]=temp
        p=p.Next
    }
    p=head
    temp=dummyNode.Next
    for p!=nil{
        if(p.Random==nil){
            temp.Random=nil
            p=p.Next
            temp=temp.Next
        }else{
            temp.Random=m[p.Random]
            p=p.Next
            temp=temp.Next
        }
    }
    return dummyNode.Next
}

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
       Node *p=head;
       unordered_map<Node*,Node*>m;
       Node *dummyNode=new Node(-1);
       Node *temp=dummyNode;
       while(p){
        temp->next=new Node(p->val);
        temp=temp->next;
        m[p]=temp;
        p=p->next;
       }
       p=head;
       temp=dummyNode->next;
       while(p){
        if(p->random==NULL){
            temp->random=NULL;
            p=p->next;
            temp=temp->next;
        }else{
            temp->random=m[p->random];
            p=p->next;
            temp=temp->next;
        }
       }
       temp=dummyNode->next;
       delete dummyNode;
       return temp;
    }
};